step-plans/step-plans-core/src/main/java/step/core/artefacts/reports/resolvedplan/ResolvedPlanBuilder.java

```diff
@@ -0,0 +1,89 @@
+package step.core.artefacts.reports.resolvedplan;
+
+import step.core.artefacts.AbstractArtefact;
+import step.core.artefacts.handlers.ArtefactHandler;
+import step.core.artefacts.handlers.ArtefactHandlerRegistry;
+import step.core.artefacts.handlers.ArtefactHashGenerator;
+import step.core.artefacts.reports.ReportNode;
+import step.core.dynamicbeans.DynamicBeanResolver;
+import step.core.dynamicbeans.DynamicJsonObjectResolver;
+import step.core.dynamicbeans.DynamicJsonValueResolver;
+import step.core.execution.ExecutionContext;
+import step.core.objectenricher.ObjectPredicate;
+import step.core.plans.Plan;
+import step.core.plans.PlanAccessor;
+import step.functions.accessor.FunctionAccessor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class ResolvedPlanBuilder {
+
+    private final PlanAccessor planAccessor;
+    private final ResolvedPlanNodeAccessor resolvedPlanNodeAccessor;
+    private final FunctionAccessor functionAccessor;
+    private final ArtefactHashGenerator artefactHashGenerator;
+    private final DynamicJsonObjectResolver dynamicJsonObjectResolver;
+    private final ArtefactHandlerRegistry artefactHandlerRegistry;
+    private final ObjectPredicate objectPredicate;
+    private final DynamicBeanResolver dynamicBeanResolver;
+
+    public ResolvedPlanBuilder(ExecutionContext executionContext) {
+        planAccessor = executionContext.getPlanAccessor();
+        resolvedPlanNodeAccessor = executionContext.require(ResolvedPlanNodeAccessor.class);
+        // functionAccessor can be null
+        functionAccessor = executionContext.get(FunctionAccessor.class);
+        artefactHandlerRegistry = executionContext.getArtefactHandlerRegistry();
+        dynamicBeanResolver = executionContext.getDynamicBeanResolver();
+        dynamicJsonObjectResolver = new DynamicJsonObjectResolver(new DynamicJsonValueResolver(executionContext.getExpressionHandler()));
+        artefactHashGenerator = new ArtefactHashGenerator();
+        objectPredicate = executionContext.getObjectPredicate();
+    }
+
+    public ResolvedPlanNode buildResolvedPlan(Plan plan, Map<String, Object> bindings) {
+        return buildTreeRecursively(null, plan.getRoot(), null, objectPredicate, bindings);
+    }
+
+    private ResolvedPlanNode buildTreeRecursively(String parentId, AbstractArtefact artefactNode, String artefactPath, ObjectPredicate objectPredicate, Map<String, Object> bindings) {
+        String artefactId = artefactNode.getId().toString();
+        String artefactHash = artefactHashGenerator.generateArtefactHash(artefactPath, artefactId);
+
+        // Resolve children
+        List<AbstractArtefact> childrenArtefacts;
```

Comment:
I'm torn: One could initialize childrenArtefacts/newArtefactPath with List.of()/null , in which case the first `else` and even `catch` implementation could go away/be empty. But I also feel that the current code, even if minimally redundant, is clearer in terms of what happens. I'll leave the decision to you, I'm fine with both.

Reply:
This code has been fully refectored in child branch 3131